### ISSUE TITLE
Potential error in loss computation?

### ISSUE THREAD

**Issue by Wuziyi616** [hyperplane-lab/Generative-3D-Part-Assembly#17]
Hi, thank you so much for open source this great work! I'm currently going through your code, and find the loss computation part a bit confusing. See [here](https://github.com/hyperplane-lab/Generative-3D-Part-Assembly/blob/1e0e671d282d24d9c95a0f0a7ae67fa923575f45/exps/Our_Method-dynamic_graph_learning/train_dynamic.py#L308-L315).

So, from my understanding, the for-loop `for iter_ind in range(conf.iter):` here is computing the MoN loss, where you do prediction multiple times and select the min loss to optimize. Then why are you doing `+=` instead of comparing the loss select the smaller one here? Say `conf.iter = 5`, then `total_loss` will be accumulated for 4 times before finally getting the `res_loss`, which is definitely larger than the loss computed in the last iter. So I think the current code will always optimize on loss in the last sampling iter

Another question is that, I think this MoN operation should be applied on each data in the batch individually. But you are doing mean [here](https://github.com/hyperplane-lab/Generative-3D-Part-Assembly/blob/1e0e671d282d24d9c95a0f0a7ae67fa923575f45/exps/Our_Method-dynamic_graph_learning/train_dynamic.py#L272). Is there a particular reason? (though I fully understand that this may only have minor effect on the final results, so taking mean is for simplicity?) Thanks in advance!

**Comment by Wuziyi616** [comment]
Ahh I see, I misunderstand some parts of the code. The implementation of MoN loss is correct :)

**Comment by Championchess** [comment]
Thanks for your interest in our work. Sorry I just saw you posted an issue. Glad to hear that you have resolved it!

**Comment by Wuziyi616** [comment]
Hi, sorry to bother again. ~~[Here](https://github.com/hyperplane-lab/Generative-3D-Part-Assembly/blob/1e0e671d282d24d9c95a0f0a7ae67fa923575f45/exps/Our_Method-dynamic_graph_learning/models/model_dynamic.py#L235-L237) you use `copy.copy`, which is shallow copy. So modifying `part_feats_copy` will also change the original `part_feats`. Is this intended?~~ This doesn't matter because we don't use this variable in the remaining part of the code.

Another question is regarding the `stop_grad` [here](https://github.com/hyperplane-lab/Generative-3D-Part-Assembly/blob/1e0e671d282d24d9c95a0f0a7ae67fa923575f45/exps/Our_Method-dynamic_graph_learning/models/model_dynamic.py#L236), which will block the gradient flow here. Is this intended? Or does it make much difference? Just feel a bit strange because intuitively, we want gradient to flow over the entire model, so that we can train the model end-to-end.

Thanks in advance!

**Comment by JialeiHuang** [comment]
We use stop_grad here to avoid some potenial gradient bug of pytorch, and indeed that operation cuts up some gradient between iterations. You can try another version without it to see the differenence if you have interest.

> Hi, sorry to bother again. ~[Here](https://github.com/hyperplane-lab/Generative-3D-Part-Assembly/blob/1e0e671d282d24d9c95a0f0a7ae67fa923575f45/exps/Our_Method-dynamic_graph_learning/models/model_dynamic.py#L235-L237) you use `copy.copy`, which is shallow copy. So modifying `part_feats_copy` will also change the original `part_feats`. Is this intended?~ This doesn't matter because we don't use this variable in the remaining part of the code.
> 
> Another question is regarding the `stop_grad` [here](https://github.com/hyperplane-lab/Generative-3D-Part-Assembly/blob/1e0e671d282d24d9c95a0f0a7ae67fa923575f45/exps/Our_Method-dynamic_graph_learning/models/model_dynamic.py#L236), which will block the gradient flow here. Is this intended? Or does it make much difference? Just feel a bit strange because intuitively, we want gradient to flow over the entire model, so that we can train the model end-to-end.
> 
> Thanks in advance!

**Comment by Wuziyi616** [comment]
Thanks, I modify the code a bit to allow gradient propagation and solve the PyTorch gradient bug, which seems to stabilize the training a little bit. But overall, that doesn't make a big difference.

**Comment by Wuziyi616** [comment]
Hi, sorry to reopen this issue. In the shape chamfer distance loss [here](https://github.com/hyperplane-lab/Generative-3D-Part-Assembly/blob/1e0e671d282d24d9c95a0f0a7ae67fa923575f45/exps/Our_Method-dynamic_graph_learning/models/model_dynamic.py#L384), I think there is an error in the implementation? In the function, you first flatten the pointcloud's `num_parts` and `num_points` dim and calculate the CD [here](https://github.com/hyperplane-lab/Generative-3D-Part-Assembly/blob/1e0e671d282d24d9c95a0f0a7ae67fa923575f45/exps/Our_Method-dynamic_graph_learning/models/model_dynamic.py#L393-L395), then, you also flatten the `valids` and multiple them with `dist1, dist2`, which is completely correct. But then, you directly take the [mean](https://github.com/hyperplane-lab/Generative-3D-Part-Assembly/blob/1e0e671d282d24d9c95a0f0a7ae67fa923575f45/exps/Our_Method-dynamic_graph_learning/models/model_dynamic.py#L399) over the `num_parts * num_points` dim, which I feel is wrong?

Let's assume there are 10 parts of a shape, each part has 1000 points. In the code we pad each shape to 20 parts right? Then `dist1, dist2` will be of shape `[batch, 20*1000]`. So taking the mean will divide the CD sum by `20*1000`, but I believe the correct number should be dividing `10*1000`. Is this true? So instead of taking the mean, we should take the sum and divide `1000 * valids.sum()`, as done in other loss (e.g. [here](https://github.com/hyperplane-lab/Generative-3D-Part-Assembly/blob/1e0e671d282d24d9c95a0f0a7ae67fa923575f45/exps/Our_Method-dynamic_graph_learning/models/model_dynamic.py#L381)).

I understand this will only affect the numbers but not the order of methods, because both your method and baselines are calculated in the same way. But since I'm implementing my own assembly codebase recently, I want to adopt the correct implementation. Also, if you can confirm this computation here is indeed wrong, then in my paper, I want to report the new numbers of baselines + your method, based on my re-implemented results (actually my re-implemented GNN assembly already outperforms the number you reported in the paper lol). I'd be grateful if you can provide any feedback regarding this issue.

**Comment by Wuziyi616** [comment]
Another question is that, when you calculate Chamfer between two `[batch, num_parts * num_points, 3]` point clouds, since the padded part point clouds are all_zeros, some valid parts (which are located near the origin, especially at the early stage of training) will be "matched" to these padded points, leading to inaccurate CD. I'd suggest a solution is to pad points with very large numbers (e.g. 999) so that no valid parts will be matched to them. Do you think this reasoning make sense?

**Comment by Wuziyi616** [comment]
I sincerely thank the authors for their kind and detailed reply in email. My questions are well addressed. The SCD loss divided over a fixed number of parts is actually an automatic hard negative sample loss weighting (so that shapes with more parts are penalized heavier), which I'm fully convinced and feel it's very novel.
